### PR TITLE
Refactor item movement

### LINKED ISSUE #11805: IMoveAction::apply(): Assertion `move_count <= count' failed
Minetest 5.4.1

experienced by Alias on IRC
https://irc.minetest.net/minetest/2021-09-12#i_5875727
Backtrace: https://pastebin.com/raw/NYw5eNBn (scroll to "Thread 4") for the relevant one.

Circumstances are unclear but it is confirmed that this only happens since #9923 (@appgurueu)
(writing this down so it isn't forgotten / someone else can look at it)

### PR DESCRIPTION
**Pull request by savilli**
Server-side item movement code had several bugs in the past (#9923, #10805, #10989), and perhaps still has (#11805).
I think the code is overcomplicated and hard to follow, which makes it difficult to find bugs.

The goal of this PR is to make the code more clear with the following steps:
1) Separate move and swap cases.
2) Remove weird patterns (swapDirections) that make code harder to understand.
3) Make it easier to follow item counters and why items won't be duplicated.

Fixes #11805 

## To do
This PR is Ready for Review.

## How to test
Use test cases from #9923, #10805, and #10989

### CODE REVIEW

**Comment by sfan5** [comment]
How do you assess the ability to add unittests for this code?

**Comment by savilli** [comment]
> How do you assess the ability to add unittests for this code?

It's not easy, but doable. Will do it in a separate PR.

**Comment by Zughy** [comment]
@savilli please rebase

**Comment by savilli** [comment]
Rebased

File: src/inventorymanager.cpp [hunk 1]
```diff
@@ -154,94 +170,257 @@ IMoveAction::IMoveAction(std::istream &is, bool somewhere) :
...
+	if (ma.to_inv.type == InventoryLocation::DETACHED)
+		can_put_count = sa->detached_inventory_AllowPut(ma, item, player);
+	else if (ma.to_inv.type == InventoryLocation::NODEMETA)
+		can_put_count = sa->nodemeta_inventory_AllowPut(ma, item, player);
+	else if (ma.to_inv.type == InventoryLocation::PLAYER)
+		can_put_count = sa->player_inventory_AllowPut(ma, item, player);
 	else
 		assert(false);
-	return dst_can_put_count;
+	return can_put_count;
```
Comment (sfan5) [inline src/inventorymanager.cpp:220]: would be a good idea imo
or is -1 a valid value too?
```suggestion
	assert(can_put_count >= 0 && can_put_count <= 0xffff);
	return can_put_count;
```

File: src/inventorymanager.cpp [hunk 2]
```diff
@@ -154,94 +170,257 @@ IMoveAction::IMoveAction(std::istream &is, bool somewhere) :
...
+			else if (can_put_count >= 0 && can_put_count < count)
+				count = can_put_count;
+		}
+	}
+
+	u16 moved_count = 0;
+	if (count > 0)
+		moved_count = list_from->moveItem(ma.from_i, list_to, ma.to_i, count);
+
+	infostream << "IMoveAction::apply(): moved " << moved_count << " from " << count << " items:"
```
Comment (sfan5) [inline src/inventorymanager.cpp:298]:
```suggestion
	infostream << "IMoveAction::apply(): moved " << moved_count << " of " << count << " items:"
```

File: src/inventorymanager.cpp [hunk 3]
```diff
@@ -154,94 +170,257 @@ IMoveAction::IMoveAction(std::istream &is, bool somewhere) :
...
+		can_take_count = sa->player_inventory_AllowMove(ma, try_take_count, player);
 	else
 		assert(false);
-	return src_can_take_count;
+	return can_take_count;
+}
+
+static std::pair<s16, bool> move(const MoveAction& ma, InventoryList *list_from, InventoryList *list_to,
+	u16 count, bool ignore_rollback, InventoryManager *mgr, ServerActiveObject *player, IGameDef *gamedef)
+{
```
Comment (sfan5) [inline src/inventorymanager.cpp:255]:
```suggestion
// returns {moved_count, can_swap}
static std::pair<s16, bool> move(const MoveAction& ma,
	InventoryList *list_from, InventoryList *list_to, u16 count, bool ignore_rollback,
	InventoryManager *mgr, ServerActiveObject *player, IGameDef *gamedef)
{
```

File: src/inventorymanager.cpp [hunk 4]
```diff
@@ -154,94 +170,257 @@ IMoveAction::IMoveAction(std::istream &is, bool somewhere) :
...
+		<< std::endl;
+
+	if (moved_count == 0)
+		return {0, count == orig_count}; // Can swap only full stack
+
+	ItemStack item = list_to->getItem(ma.to_i);
+	item.count = moved_count;
+
+	if (inf_from) {
+		ItemStack& item = list_from->getItem(ma.from_i);
```
Comment (sfan5) [inline src/inventorymanager.cpp:315]: please don't shadow `item`, that's confusing

File: src/inventorymanager.cpp [hunk 5]
```diff
@@ -154,94 +170,257 @@ IMoveAction::IMoveAction(std::istream &is, bool somewhere) :
...
+		<< " i=" << ma.to_i
+		<< std::endl;
+
+	if (moved_count == 0)
+		return {0, count == orig_count}; // Can swap only full stack
+
+	ItemStack item = list_to->getItem(ma.to_i);
+	item.count = moved_count;
+
+	if (inf_from) {
```
Comment (sfan5) [inline src/inventorymanager.cpp:314]:
```suggestion
	// If source or dest are meant to be infinite, undo the change
	if (inf_from) {
```

File: src/inventorymanager.cpp [hunk 6]
```diff
@@ -278,248 +457,73 @@ void IMoveAction::apply(InventoryManager *mgr, ServerActiveObject *player, IGame
...
-		move_somewhere = false;
+	if (from_i < 0 || list_from->getSize() <= (u32) from_i) {
+		infostream << "IMoveAction::apply(): FAIL: source index out of bounds: "
+			<< "size of from_list=\"" << list_from->getSize() << "\""
+			<< ", from_index=\"" << from_i << "\"" << std::endl;
+		return;
+	}
 
+	// Do not handle rollback if both inventories are that of the same player
+	bool ignore_rollback = from_inv.type == InventoryLocation::PLAYER && from_inv == to_inv;
```
Comment (sfan5) [inline src/inventorymanager.cpp:468]: I don't think rollback can work for detached inventories either

**Review by sfan5 — COMMENTED** [review]
Looks good, thanks for cleaning this up.
I'll try to give it a quick test later and then approve

File: src/inventorymanager.cpp (hunk 3 above)
Comment (SmallJoker) [reply src/inventorymanager.cpp:255]: Why is this not a member function of `MoveAction`? That object holds the inventory list data and move count. It would shorten this function call quite a bit.

File: src/inventorymanager.cpp [hunk 7]
```diff
@@ -154,94 +170,257 @@ IMoveAction::IMoveAction(std::istream &is, bool somewhere) :
...
 
-void IMoveAction::swapDirections()
-{
-	std::swap(from_inv, to_inv);
-	std::swap(from_list, to_list);
-	std::swap(from_i, to_i);
-}
-
-void IMoveAction::onPutAndOnTake(const ItemStack &src_item, ServerActiveObject *player) const
+static void onPutAndOnTake(const MoveAction& ma, const ItemStack &item, ServerActiveObject *player)
```
Comment (SmallJoker) [inline src/inventorymanager.cpp:173]: What's the goal of making these functions static? They clearly belong to the `MoveAction` class.

**Comment by sfan5** [comment]
Test result: :-1: Taking an item from the MTG creative inventory actually removes it.

**Comment by Zughy** [comment]
@savilli please take care of reviews

File: src/inventorymanager.cpp (hunk 1 above)
Comment (savilli) [reply src/inventorymanager.cpp:220]: -1 is a valid value too, https://github.com/minetest/minetest/blob/8fded9d990d1e1142b630b24c96f1a2c959fef36/doc/lua_api.txt#L8640

File: src/inventorymanager.cpp (hunk 2 above)
Comment (savilli) [reply src/inventorymanager.cpp]: Fixed

File: src/inventorymanager.cpp (hunk 3 above)
Comment (savilli) [reply src/inventorymanager.cpp]: Fixed

File: src/inventorymanager.cpp (hunk 4 above)
Comment (savilli) [reply src/inventorymanager.cpp]: Fixed

File: src/inventorymanager.cpp (hunk 5 above)
Comment (savilli) [reply src/inventorymanager.cpp]: Added

File: src/inventorymanager.cpp (hunk 6 above)
Comment (savilli) [reply src/inventorymanager.cpp:456]: It probably doesn't, but it was like that before, and I don't think it's worth changing in this PR.

File: src/inventorymanager.cpp (hunk 7 above)
Comment (savilli) [reply src/inventorymanager.cpp]: I don't remember XD.
Moved to class functions again.

**Comment by savilli** [comment]
> Test result: -1 Taking an item from the MTG creative inventory actually removes it.

Fixed. Need to restore the whole item if it's infinite, not only the counter.

File: src/inventorymanager.cpp (hunk 1 above)
Comment (sfan5) [reply src/inventorymanager.cpp:220]: Should we check that it's < 65535 then?

Comment (savilli) [reply src/inventorymanager.cpp:220]: What is your goal here? It makes little sense for a mod to return a count > 65535, but it doesn't break anything. The code still would work as expected.

Comment (sfan5) [reply src/inventorymanager.cpp:220]: The intent here is to check for invalid situations early, preventing potential problems.

Comment (savilli) [reply src/inventorymanager.cpp:220]: Okay, I don't mind adding it

**Comment by Zughy** [comment]
@savilli please rebase
